### PR TITLE
(PUP-7114) Re-vendor semantic_puppet gem

### PR DESCRIPTION
**Pull request by Iristyle**
- As part of PUP-4742 the dependency on semantic_puppet was moved to
   a gem.  The sequence of events leading up to its conclusion is
   rather thorny, but ultimately code landed in
   c111fb76b948d547fda7afa901802ff2e8b5a623

 - Because puppet-server was not also shipping the semantic_puppet gem
   this necessitated a strict conflict in the puppet-agent package
   on puppet-server < 2.7.2 and pe-puppet-server < 2017.0.1.3

   This has had quite a bit of fallout, including unintended upgrades
   of the puppet-server package, which is extermely undesirable in a
   production environment.

 - Until Puppet 5 is released, which will allow for more latitude of
   breaking changes and dependency requirements, it has been decided
   that semantic_puppet should be temporarily re-vendored into the
   Puppet repository.  Then there is no need to match the gem in both
   puppet-server and puppet-agent.  Instead, the code from this
   repository is loaded.

 - References to the semantic_puppet gem code have been removed from
   the Gemfile, project_data.yaml and the Puppet .gemspec.

   The code from semantic_puppet 0.1.4 has been directly copied into
   lib/semantic_puppet

### CODE REVIEW

**Comment by MosesMendoza** [comment]
It seems like we can't put this in `lib/puppet/vendor` with the other vendored deps because we need it loaded earlier? Unfortunate - that seems like it might have been more intuitive there.

File: lib/semantic_puppet.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+require 'gettext-setup'
+
+module SemanticPuppet
+  GettextSetup.initialize(File.absolute_path('../locales', File.dirname(__FILE__)))
```
Comment (MosesMendoza) [inline lib/semantic_puppet.rb]: Is the plan still to avoid a hard dependency on `gettext-setup`? If so is that change intended for a separate PR?

Comment (Iristyle) [inline lib/semantic_puppet.rb]: Yeah, I was thinking we'd do a roundup on all the gettext related work in a separate PR, but in this PR would also work if anyone feels strongly.

**Comment by Iristyle** [comment]
@MosesMendoza we could place in `lib/puppet/vendor` but then would have to change any existing `require` statements.  It can be done, but felt simpler to do a drop-in replacement, since we know it will be short-lived.

**Comment by MosesMendoza** [comment]
Sounds good. Care to add a short-lived exclude .rubocop.yml for semantic_puppet? We already exclude other things in vendor

File: lib/semantic_puppet.rb (hunk 1 above)
Comment (MosesMendoza) [inline lib/semantic_puppet.rb]: 👍

**Comment by Magisus** [comment]
👍

**Comment by Iristyle** [comment]
For historical purposes, there's a long history here of 6 PRs back to February 2016:

* #4708 
* #4977 
* #5008 
* #5094 
* #5399 
* #5400
